File: WitchcraftApp/src/test/java/com/tarbadev/witchcraft/rest/RecipesRestControllerTest.java
package com.tarbadev.witchcraft.rest;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.tarbadev.witchcraft.TestResources;
import com.tarbadev.witchcraft.domain.entity.Ingredient;
import com.tarbadev.witchcraft.domain.entity.Recipe;
import com.tarbadev.witchcraft.domain.entity.Step;
import com.tarbadev.witchcraft.domain.usecase.*;
import com.tarbadev.witchcraft.web.IngredientModifyForm;
import com.tarbadev.witchcraft.web.RecipeFormRequest;
import com.tarbadev.witchcraft.web.RecipeModifyForm;
import com.tarbadev.witchcraft.web.StepModifyForm;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.Arrays.asList;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.core.Is.is;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@RunWith(SpringRunner.class)
@SpringBootTest
@AutoConfigureMockMvc
public class RecipesRestControllerTest {
  @Autowired private MockMvc mvc;
  @Autowired private TestResources testResources;
  @Autowired private RecipeCatalogUseCase recipeCatalogUseCase;
  @Autowired private GetRecipeUseCase getRecipeUseCase;
  @Autowired private DeleteRecipeUseCase deleteRecipeUseCase;
  @Autowired private DoesRecipeExistUseCase doesRecipeExistUseCase;
  @Autowired private SetFavoriteRecipeUseCase setFavoriteRecipeUseCase;
  @Autowired private GetRecipeDetailsFromUrlUseCase getRecipeDetailsFromUrlUseCase;
  @Autowired private SaveRecipeUseCase saveRecipeUseCase;
  @Autowired private GetRecipeDetailsFromFormUseCase getRecipeDetailsFromFormUseCase;
  @Autowired private GetFavoriteRecipesUseCase getFavoriteRecipesUseCase;
  @Autowired private LastAddedRecipesUseCase lastAddedRecipesUseCase;

  @Before
  public void setUp() {
    Mockito.reset(
        recipeCatalogUseCase,
        getRecipeUseCase,
        setFavoriteRecipeUseCase,
        deleteRecipeUseCase,
        getFavoriteRecipesUseCase,
        lastAddedRecipesUseCase
    );
  }

  @Test
  public void list() throws Exception {
    String recipe1Name = "Test";
    String recipe1ImageUrl = "exampleImageUrl";
    Integer recipe1Id = 10;
    String recipe1Url = "/recipes/" + recipe1Id;
    String recipe2Name = "New Lasagna";

    Recipe recipe1 = Recipe.builder()
        .id(recipe1Id)
        .name(recipe1Name)
        .imgUrl(recipe1ImageUrl)
        .build();

    List<Recipe> recipes = asList(
        recipe1,
        Recipe.builder().name(recipe2Name).build()
    );

    given(recipeCatalogUseCase.execute()).willReturn(recipes);

    mvc.perform(get("/api/recipes"))
        .andExpect(status().isOk())
        .andExpect(jsonPath("$.recipes", hasSize(2)))
        .andExpect(jsonPath("$.recipes[0].id", is(recipe1Id)))
        .andExpect(jsonPath("$.recipes[0].name", is(recipe1Name)))
        .andExpect(jsonPath("$.recipes[0].imgUrl", is(recipe1ImageUrl)))
        .andExpect(jsonPath("$.recipes[0].url", is(recipe1Url)))
        .andExpect(jsonPath("$.recipes[1].name", is(recipe2Name)));
  }

  @Test
  public void show() throws Exception {
    Recipe recipe = Recipe.builder()
        .id(33)
        .name("Test")
        .url("url")
        .build();

    given(getRecipeUseCase.execute(recipe.getId())).willReturn(recipe);

    mvc.perform(get("/api/recipes/" + recipe.getId()))
        .andExpect(status().isOk())
        .andExpect(jsonPath("$.id", is(recipe.getId())))
        .andExpect(jsonPath("$.url", is(recipe.getUrl())))
        .andExpect(jsonPath("$.name", is(recipe.getName())));
  }

  @Test
  public void setFavorite() throws Exception {
    int id = 32;
    Boolean favorite = true;

    mvc.perform(patch(String.format("/api/recipes/%s", id))
        .contentType(MediaType.APPLICATION_JSON)
        .content(String.format("{ \"favorite\": %s }", favorite.toString()))
    )
        .andExpect(status().is(HttpStatus.NO_CONTENT.value()));

    verify(setFavoriteRecipeUseCase).execute(id, favorite);
  }

  @Test
  public void delete() throws Exception {
    int id = 32;

    given(doesRecipeExistUseCase.execute(id)).willReturn(true);

    mvc.perform(MockMvcRequestBuilders.delete(String.format("/api/recipes/%s", id)))
        .andExpect(status().is(HttpStatus.NO_CONTENT.value()));

    verify(deleteRecipeUseCase).execute(id);
  }

  @Test
  public void delete_returnsNotFound_whenRecipeIsNotFound() throws Exception {
    int id = 32;

    given(doesRecipeExistUseCase.execute(id)).willReturn(false);

    mvc.perform(MockMvcRequestBuilders.delete(String.format("/api/recipes/%s", id)))
        .andExpect(status().isNotFound());

    verify(deleteRecipeUseCase, never()).execute(id);
  }

  @Test
  public void importFromUrl() throws Exception {
    Recipe recipe = testResources.getRecipe();
    RecipeFormRequest recipeFormRequest = RecipeFormRequest.builder().url(recipe.getOriginUrl()).build();

    given(getRecipeDetailsFromUrlUseCase.execute(recipe.getOriginUrl())).willReturn(recipe);
    given(saveRecipeUseCase.execute(recipe)).willReturn(recipe);

    mvc.perform(post("/api/recipes/importFromUrl")
        .contentType(MediaType.APPLICATION_JSON)
        .content(new ObjectMapper().writeValueAsString(recipeFormRequest))
    )
        .andExpect(content().json(new ObjectMapper().writeValueAsString(recipe)));

    verify(getRecipeDetailsFromUrlUseCase).execute(recipe.getOriginUrl());
    verify(saveRecipeUseCase).execute(recipe);
  }

  @Test
  public void importFromForm() throws Exception {
    RecipeFormRequest recipeFormRequest = RecipeFormRequest.builder()
        .name("Some recipe name")
        .url("http://some/url/of/recipe")
        .imageUrl("http://some/url/of/recipe.png")
        .ingredients(String.join("\n"
            , "10 tbsp sugar"
            , "1/2 cup olive oil"
            , "1 lemon"
        ))
        .steps(String.join("\n"
            , "Add ingredients and stir"
            , "Serve on each plate"
        ))
        .build();
    Recipe recipe = Recipe.builder()
        .name(recipeFormRequest.getName())
        .originUrl(recipeFormRequest.getUrl())
        .imgUrl(recipeFormRequest.getImageUrl())
        .ingredients(Arrays.stream(recipeFormRequest.getIngredients().split("\n")).map(ingredient -> Ingredient.builder().name(ingredient).build()).collect(Collectors.toList()))
        .steps(Arrays.stream(recipeFormRequest.getSteps().split("\n")).map(step -> Step.builder().name(step).build()).collect(Collectors.toList()))
        .build();

    given(getRecipeDetailsFromFormUseCase.execute(
        recipeFormRequest.getName(),
        recipeFormRequest.getUrl(),
        recipeFormRequest.getIngredients(),
        recipeFormRequest.getSteps(),
        recipeFormRequest.getImageUrl()
    )).willReturn(recipe);
    given(saveRecipeUseCase.execute(recipe)).willReturn(recipe);

    mvc.perform(post("/api/recipes/importFromForm")
        .contentType(MediaType.APPLICATION_JSON)
        .content(new ObjectMapper().writeValueAsString(recipeFormRequest))
    )
        .andExpect(content().json(new ObjectMapper().writeValueAsString(recipe)));
  }

  @Test
  public void update() throws Exception {
    RecipeModifyForm recipeModifyForm = RecipeModifyForm.builder()
        .id(12)
        .name("Recipe Name")
        .url("http://fake/url")
        .imgUrl("http://fake/url.png")
        .favorite(true)
        .ingredients(Collections.singletonList(
            IngredientModifyForm.builder().name("Ingredient").build()
        ))
        .steps(Collections.singletonList(StepModifyForm.builder().name("First Step").build()))
        .build();

    Recipe recipe = Recipe.builder()
        .id(recipeModifyForm.getId())
        .name(recipeModifyForm.getName())
        .originUrl(recipeModifyForm.getUrl())
        .imgUrl(recipeModifyForm.getImgUrl())
        .favorite(recipeModifyForm.getFavorite())
        .ingredients(recipeModifyForm.getIngredients().stream()
            .map(ingredientModifyForm -> Ingredient.builder()
                .id(ingredientModifyForm.getId())
                .name(ingredientModifyForm.getName())
                .unit(ingredientModifyForm.getUnit())
                .quantity(ingredientModifyForm.getQuantity())
                .build()
            ).collect(Collectors.toList())
        )
        .steps(recipeModifyForm.getSteps().stream()
            .map(stepModifyForm -> Step.builder()
                .id(stepModifyForm.getId())
                .name(stepModifyForm.getName())
                .build()
            )
            .collect(Collectors.toList())
        )
        .build();

    given(saveRecipeUseCase.execute(recipe)).willReturn(recipe);

    mvc.perform(put("/api/recipes/" + recipeModifyForm.getId() + "/update")
        .contentType(MediaType.APPLICATION_JSON)
        .content(new ObjectMapper().writeValueAsString(recipeModifyForm))
    )
        .andExpect(status().isOk())
        .andExpect(content().json(new ObjectMapper().writeValueAsString(recipe)));
  }

  @Test
  public void favorites() throws Exception {
    List<Recipe> recipes = asList(
        Recipe.builder().build(),
        Recipe.builder().build(),
        Recipe.builder().build()
    );

    given(getFavoriteRecipesUseCase.execute()).willReturn(recipes);

    mvc.perform(get("/api/recipes/favorites"))
        .andExpect(status().isOk())
        .andExpect(content().json(new ObjectMapper().writeValueAsString(recipes)));
  }

  @Test
  public void latest() throws Exception {
    List<Recipe> recipes = asList(
        Recipe.builder().build(),
        Recipe.builder().build(),
        Recipe.builder().build()
    );

    given(lastAddedRecipesUseCase.execute()).willReturn(recipes);

    mvc.perform(get("/api/recipes/latest"))
        .andExpect(status().isOk())
        .andExpect(content().json(new ObjectMapper().writeValueAsString(recipes)));
  }
}

File: WitchcraftApp/src/test/kotlin/com/tarbadev/witchcraft/recipes/rest/RecipesRestControllerTest.kt
package com.tarbadev.witchcraft.recipes.rest

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import com.tarbadev.witchcraft.converter.unit
import com.tarbadev.witchcraft.recipes.domain.entity.Ingredient
import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import com.tarbadev.witchcraft.recipes.domain.entity.Step
import com.tarbadev.witchcraft.recipes.domain.entity.StepNote
import com.tarbadev.witchcraft.recipes.domain.usecase.*
import com.tarbadev.witchcraft.recipes.rest.entity.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

@ExtendWith(SpringExtension::class)
@WebMvcTest(RecipesRestController::class)
class RecipesRestControllerTest(
    @Autowired private val mockMvc: MockMvc
) {
  @MockBean
  private lateinit var recipeCatalogUseCase: RecipeCatalogUseCase
  @MockBean
  private lateinit var getRecipeUseCase: GetRecipeUseCase
  @MockBean
  private lateinit var deleteRecipeUseCase: DeleteRecipeUseCase
  @MockBean
  private lateinit var doesRecipeExistUseCase: DoesRecipeExistUseCase
  @MockBean
  private lateinit var setFavoriteRecipeUseCase: SetFavoriteRecipeUseCase
  @MockBean
  private lateinit var saveRecipeUseCase: SaveRecipeUseCase
  @MockBean
  private lateinit var getFavoriteRecipesUseCase: GetFavoriteRecipesUseCase
  @MockBean
  private lateinit var lastAddedRecipesUseCase: LastAddedRecipesUseCase
  @MockBean
  private lateinit var editStepNoteUseCase: EditStepNoteUseCase
  @MockBean
  private lateinit var updatePortionsUseCase: UpdatePortionsUseCase

  @Test
  fun list() {
    val recipe1Name = "Test"
    val recipe1ImageUrl = "exampleImageUrl"
    val recipe1Id = 10
    val recipe2Name = "New Lasagna"

    val recipe1 = Recipe(
        id = recipe1Id,
        name = recipe1Name,
        imgUrl = recipe1ImageUrl
    )

    val recipes = listOf(
        recipe1,
        Recipe(name = recipe2Name)
    )

    whenever(recipeCatalogUseCase.execute()).thenReturn(recipes)

    mockMvc.perform(get("/api/recipes"))
        .andExpect(status().isOk)
        .andExpect(header().string(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(RecipeListResponse.fromRecipeList(recipes))))
  }

  @Test
  fun show() {
    val recipe = Recipe(
        id = 33,
        name = "Test"
    )

    whenever(getRecipeUseCase.execute(recipe.id)).thenReturn(recipe)

    mockMvc.perform(get("/api/recipes/${recipe.id}"))
        .andExpect(status().isOk)
        .andExpect(header().string(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(RecipeResponse.fromRecipe(recipe))))
  }

  @Test
  fun setFavorite() {
    val id = 32
    val favorite = true
    val recipe = Recipe(id = id, favorite = favorite)

    whenever(setFavoriteRecipeUseCase.execute(id, favorite)).thenReturn(recipe)

    mockMvc.perform(patch("/api/recipes/${recipe.id}")
        .contentType(MediaType.APPLICATION_JSON_VALUE)
        .content(jacksonObjectMapper().writeValueAsString(SetFavoriteRequest(favorite)))
    )
        .andExpect(status().isOk)
        .andExpect(header().string(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(RecipeResponse.fromRecipe(recipe))))
  }

  @Test
  fun delete() {
    val id = 32

    whenever(doesRecipeExistUseCase.execute(id)).thenReturn(true)

    mockMvc.perform(delete("/api/recipes/$id"))
        .andExpect(status().isNoContent)

    verify(deleteRecipeUseCase).execute(id)
  }

  @Test
  fun delete_returnsNotFound_whenRecipeIsNotFound() {
    val id = 32

    whenever(doesRecipeExistUseCase.execute(id)).thenReturn(false)

    mockMvc.perform(delete("/api/recipes/$id"))
        .andExpect(status().isNotFound)

    verify(deleteRecipeUseCase, never()).execute(id)
  }

  @Test
  fun update() {
    val recipeModifyRequest = RecipeModifyRequest(
        id = 12,
        name = "Recipe Name",
        url = "http://fake/url",
        imgUrl = "http://fake/url.png",
        favorite = true,
        ingredients = listOf(IngredientModifyRequest(name = "Ingredient")),
        steps = listOf(StepModifyRequest(name = "First Step"))
    )

    val recipe = Recipe(
        id = recipeModifyRequest.id,
        name = recipeModifyRequest.name,
        originUrl = recipeModifyRequest.url,
        imgUrl = recipeModifyRequest.imgUrl,
        favorite = recipeModifyRequest.favorite,
        ingredients = recipeModifyRequest.ingredients
            .map { ingredientModifyRequest ->
              Ingredient(
                  id = ingredientModifyRequest.id,
                  name = ingredientModifyRequest.name,
                  quantity = 0.unit
              )
            },
        steps = recipeModifyRequest.steps
            .map { stepModifyRequest ->
              Step(
                  id = stepModifyRequest.id,
                  name = stepModifyRequest.name
              )
            },
        isArchived = false
    )

    whenever(saveRecipeUseCase.execute(recipe)).thenReturn(recipe)

    mockMvc.perform(put("/api/recipes/${recipeModifyRequest.id}/update")
        .contentType(MediaType.APPLICATION_JSON_VALUE)
        .content(jacksonObjectMapper().writeValueAsString(recipeModifyRequest))
    )
        .andExpect(status().isOk)

    verify(saveRecipeUseCase).execute(recipe)
  }

  @Test
  fun favorites() {
    val recipe1 = Recipe()
    val recipe2 = Recipe()
    val recipe3 = Recipe()

    val recipes = listOf(recipe1, recipe2, recipe3)

    val recipesResponse = listOf(
        RecipeResponse.fromRecipe(recipe1),
        RecipeResponse.fromRecipe(recipe2),
        RecipeResponse.fromRecipe(recipe3)
    )

    whenever(getFavoriteRecipesUseCase.execute()).thenReturn(recipes)

    mockMvc.perform(get("/api/recipes/favorites"))
        .andExpect(status().isOk)
        .andExpect(header().string(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(recipesResponse)))
  }

  @Test
  fun latest() {
    val recipe1 = Recipe()
    val recipe2 = Recipe()
    val recipe3 = Recipe()

    val recipes = listOf(recipe1, recipe2, recipe3)

    val recipesResponse = listOf(
        RecipeResponse.fromRecipe(recipe1),
        RecipeResponse.fromRecipe(recipe2),
        RecipeResponse.fromRecipe(recipe3)
    )

    whenever(lastAddedRecipesUseCase.execute()).thenReturn(recipes)

    mockMvc.perform(get("/api/recipes/latest"))
        .andExpect(status().isOk)
        .andExpect(header().string(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(recipesResponse)))
  }

  @Test
  fun addStepNote() {
    val editStepNoteRequest = EditStepNoteRequest("Some step note")
    val newStep = Step(id = 78, name = "Some step name", note = StepNote(id = 98, comment = editStepNoteRequest.note))
    val stepResponse = StepResponse.fromStep(newStep)

    whenever(editStepNoteUseCase.execute(stepResponse.id, editStepNoteRequest.note)).thenReturn(newStep)

    mockMvc.perform(post("/api/recipes/45/steps/78")
        .contentType(MediaType.APPLICATION_JSON)
        .content(jacksonObjectMapper().writeValueAsString(editStepNoteRequest))
    )
        .andExpect(status().isOk)
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(stepResponse)))
  }

  @Test
  fun updatePortions() {
    val updatePortionsRequest = UpdatePortionsRequest(4)
    val newRecipe = Recipe(id = 45, portions = 4)
    val recipeResponse = RecipeResponse.fromRecipe(newRecipe)

    whenever(updatePortionsUseCase.execute(recipeResponse.id, updatePortionsRequest.portions)).thenReturn(newRecipe)

    mockMvc.perform(put("/api/recipes/45/portions")
        .contentType(MediaType.APPLICATION_JSON)
        .content(jacksonObjectMapper().writeValueAsString(updatePortionsRequest))
    )
        .andExpect(status().isOk)
        .andExpect(content().json(jacksonObjectMapper().writeValueAsString(recipeResponse)))
  }
}